File: Python3.x/295-Find Median from Data Stream.py
class MedianFinder:
    
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.lst = []

    def addNum(self, num: int) -> None:
        bisect.insort(self.lst, num)

    def findMedian(self) -> float:
        len_lst = len(self.lst)
        if len_lst % 2 == 1:
            return self.lst[len_lst // 2]
        else:
            return (self.lst[len_lst // 2 - 1] + self.lst[len_lst // 2]) / 2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()




# binary search insert
class MedianFinder:
    
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.lst = []

    def addNum(self, num: int) -> None:
        if len(self.lst) == 0 or self.lst[-1] < num:
            self.lst.append(num)
            return
        left, right = 0, len(self.lst) - 1
        while left <= right:
            mid = (left + right) // 2
            if num < self.lst[mid]:
                right = mid - 1
            else:
                left = mid + 1
        self.lst.insert(left, num)

    def findMedian(self) -> float:
        len_lst = len(self.lst)
        if len_lst % 2 == 1:
            return self.lst[len_lst // 2]
        else:
            return (self.lst[len_lst // 2 - 1] + self.lst[len_lst // 2]) / 2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: Algorithms/C++/295-Find_Median_from_Data_Stream.cpp
class MedianFinder {
private:
    vector<int> vec;
public:
    /** initialize your data structure here. */
    MedianFinder() {
        
    }
    
    void addNum(int num) {
        if (vec.size() == 0) {
            vec.push_back(num);
        } else {
            int l = 0, r = vec.size() - 1;
            while (l <= r) {
                int mid = (l + r) / 2;
                if (num < vec[mid])
                    r = mid - 1;
                else
                    l = mid + 1;
            }
            vec.insert(vec.begin() + l, num);
        }
    }
    
    double findMedian() {
        int vec_size = vec.size();
        if (vec_size & 1)
            return vec[vec_size / 2];
        else
            return (vec[vec_size / 2 - 1] + vec[vec_size / 2]) * 0.5;
    }
};